workitems/tokenlist/l2-token-list-v1.0-psd01.md

```diff
@@ -236,29 +238,43 @@ The tokens data element is a composite which MUST minimally contain the followin
 
 Note, that the detailed definition of the data elements in [[R2]](#r2) along with descriptions and examples are given in the schema itself below.
 
+[[R2]](#r2) testability: See suggested test fixtures for this documents' data schema below.
+
 #### **[D1]**
 All other data elements of the schema SHOULD be included in a representation of a canonical token list.
 
+[[D1]](#d1) testability: See suggested test fixtures for this documents' data schema below.
+
 #### **[CR1]>[D1]**
-if the extension data elements is used, the following data elements MUST be present in the schema representation:
+If the extension data elements is used, the following data elements MUST be present in the schema representation:
 - rootChainId
 - rootChainURI
 - rootAddress
 
 Note, that the detailed definition of the data elements in [[D1]](#d1) and [[CR1]>[D1]](#cr1d1) along with descriptions and examples are given in the schema itself below.
 
+[[CR1]>[D1]](#cr1d1) testability: See suggested test fixtures for this documents' data schema below.
+
 #### **[R3]**
 All properties in the schema identified in the description to be a Universal Resource Identifier (URI) MUST follow in their semantics [RFC 3986](#rfc3986).
 
+[[R3]](#r3) testability: As an approved standard, all requirements for [RFC 3986](#rfc3986) are testable.
+
 #### **[R4]**
 The `chainId` property MUST follow [EIP-155](#eip155) standard.
 
+[[R4]](#r4) testability: As an approved standard, all requirements for [EIP-155](#eip155) are testable.
```

Comment:
More specifically, EIP-155 isn't a standard for a unique chainID - it's for a way to use such a chainID.

I suspect a better approach would be to require that the chainID is registered in https://github.com/ethereum-lists/chains according to the requirements of that repo, including that it is unique (at least among chains registered there). It's a sadly centralised approach to achieving decentralisation... :(

Reply:
addressed, see new commit @chaals

Reply:
@Therecanbeonlyone1969 seems we were working at the same time. As noted again in my second comment with more explanation, and in #36, I think the problem here is that conformance to EIP-155 isn't actually what the spec should require. So https://github.com/eea-oasis/L2/commit/2664238be5c00a16df532f14af585131c7d9e1c2 doesn't address this comment :S

Reply:
@chaals it is probably not a good idea to put a possibly impermanent URL into a requirement (EIP-155 actually has external links). While not perfect, implementers will understand how to use a chainId. EIP-155 implies uniqueness otherwise there is no replay protection.

Reply:
EIP implies uniqueness, since otherwise it doesn't work, but there is no requirement in EIP for uniqueness. Since the best idea EIP offers on how to have uniqueness is through registering things that can be looked up, its pointers to those addresses are the most valuable part of the whole spec for anyone (mis-)using it as a reference to the need for unique chain identifiers.

to require uniqueness of identifiers, and I believe that's a pretty critical piece of the work here, I suggest we explicitly do so in this spec, and describe how to achieve that.

Which will incidentally then make this spec a much better reference than EIP-155, solving the problem that lots of people try to use it as if it required something it doesn't.

Reply:
@chaals I totally hear you, and agree with you in principle. Taking a step back, what you are suggesting is a significant step since it impacts current implementations of L1s and L2s = Hornets Nest => to be avoided atm. Also, this spec is a schema spec, not a spec for chainId. 

Also, it is not in the remit of the L2 WG to right all the issues in the Ethereum ecosystem not unique to  L2. Since we want to make this spec an EIP, and do not want to boil the ocean at the same time, it is not a good idea to tackle chainId uniqueness beyond what is implied by EIP-155.

Reply:
@chaals good to close this out?

Reply:
No. I agree the aim isn't to solve all problems, but I disagree that it is reasonable for this specification not to solve the problem of ensuring that a chainID is unique.

The abstract says

> The document describes the minimal set of business and technical prerequisites, functional and non-functional requirements for a token list that when implemented ensures that two or more Layer 1, Layer 2, or Sidechains can identify tokens from different Layer 1, Layer 2, or Sidechains.

I believe that accurately identifying a chain is a technical prererequisite to identify tokens on that chain. As far as I can tell, the proposal is to rely on the fact that EIP assumes that there is a way to do that, explicitly ignoring that it doesn't make it a requirement, and deliberately not referring to the hints in the EIP about a mechanism that might work if it's used properly.

That seems a particularly bad approach to me, in particular because it seems to reject the idea that this specification should achieve what it states is its purpose.

Reply:
@chaals 
- this standard is about a token list not defining a `chainId`. Such a definition is beyond the scope of this particular standard. 
- the `chainId` is not the only piece of data used to identify a token within the token list. Therefore, other data is in fact necessary to validate the existence of the token itself. In other words, `chainId` uniqueness may be helpful but not required.
- not trying to be facetious but the abstract says `can identify` not `can uniquely identify`.
- we will discuss this in the WG during our PR discussion on 1/25, and resolve

Reply:
First, an apology because I think I haven't been clear, and because at some point I can see that my thinking got too far into one detail and my train of thought lost the entirety of the spec.

I think I understand the goal of the spec. But if you can't uniquely identify the chain the token is on, I don't think that the spec meets that goal - because it becomes possible to match on another chain. 

I'm thinking of https://banteg.mirror.xyz/iZAsBNL3j_5NIAY2Erav1r7Q4ecc7SC76AfMjyScs34 (via https://ventral.digital/posts/2022/12/15/ethereum-smart-contract-auditors-2022-rewind#transaction-replay-attack is where I saw it). It's not the same, but it "rhymes" at least in my mind.

I'm also meaning "a unique identifier for the chain" rather than the `chainID` property (sorry, I should have been much clearer about that) - I believe you are stating that the set of requirements (not just `chainID`) does provide that

It seems to me that unless you enforce both EIP-155 replay protection *and* having a unique identifier for the chain (which is assumed by EIP-155 rather than required) you leave open a vulnerability.

Reading the spec again the requirement for a `chainURI` pointing to a (verifiable) genesis block helps, because it reduces the scope for vulnerabilities to a fork, which is a much smaller attack surface.

It still isn't obvious to me that the requirements overall mean that attack surface has been eliminated (nor even that `genesisBlockHash` that makes `chainURI` properly verifiable is **required**), so I'm still concerned.

(That said, it is possible I'm also wrong, and there *isn't* a vulnerability here).

Reply:
@chaals The WG during its 1/25 meeting discussed the uniqueness point you raised. It was decided that we will not add a uniqueness requirement. A security advisory was added in the security considerations that implementers of the canonical token list standard must ensure the correctness and security of the data they display. Happy to discuss this further on a topic and possibly create an EIP for `chainId` uniqueness. Will resolve this conversation for now.